// react-hover/dist/lib/Trigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Trigger = function (_Component) {
  _inherits(Trigger, _Component);

  function Trigger(props) {
    _classCallCheck(this, Trigger);

    var _this = _possibleConstructorReturn(this, (Trigger.__proto__ || Object.getPrototypeOf(Trigger)).call(this, props));

    _this.state = {
      styles: {}
    };
    return _this;
  }

  _createClass(Trigger, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var childStyles = this.refs.triggerContainer.children[0].style;
      this.setState({
        styles: {
          width: childStyles.width,
          height: childStyles.height,
          margin: childStyles.margin
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var styles = this.state.styles;

      return _react2.default.createElement(
        'div',
        {
          onMouseOver: this.onMouseOver.bind(this),
          onMouseOut: this.onMouseOut.bind(this),
          onMouseMove: this.onMouseMove.bind(this),
          onTouchStart: this.onTouchStart.bind(this),
          onTouchEnd: this.onTouchEnd.bind(this),
          ref: 'triggerContainer',
          style: styles
        },
        this.props.children.props.children
      );
    }
  }, {
    key: 'onMouseOver',
    value: function onMouseOver() {
      var setVisibility = this.props.children.props.setVisibility;

      setVisibility(true);
    }
  }, {
    key: 'onMouseOut',
    value: function onMouseOut() {
      var setVisibility = this.props.children.props.setVisibility;

      setVisibility(false);
    }
  }, {
    key: 'onMouseMove',
    value: function onMouseMove(e) {
      var getCursorPos = this.props.children.props.getCursorPos;

      getCursorPos(e);
    }
  }, {
    key: 'onTouchStart',
    value: function onTouchStart() {
      var setVisibility = this.props.children.props.setVisibility;

      setVisibility(true);
    }
  }, {
    key: 'onTouchEnd',
    value: function onTouchEnd() {
      var setVisibility = this.props.children.props.setVisibility;

      setVisibility(false);
    }
  }]);

  return Trigger;
}(_react.Component);

Trigger.propTypes = {
  type: _propTypes2.default.string,
  children: _propTypes2.default.object,
  setVisibility: _propTypes2.default.func,
  getCursorPos: _propTypes2.default.func
};
exports.default = Trigger;
