// react-hover/dist/ReactHover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Hover = require('./lib/Hover');

var _Hover2 = _interopRequireDefault(_Hover);

var _Trigger = require('./lib/Trigger');

var _Trigger2 = _interopRequireDefault(_Trigger);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ReactHover = function (_Component) {
  _inherits(ReactHover, _Component);

  function ReactHover(props) {
    _classCallCheck(this, ReactHover);

    var _this = _possibleConstructorReturn(this, (ReactHover.__proto__ || Object.getPrototypeOf(ReactHover)).call(this, props));

    _this.state = {
      children: _propTypes2.default.object,
      hoverComponentStyle: {
        display: 'none',
        position: 'absolute'
      }
    };
    return _this;
  }

  _createClass(ReactHover, [{
    key: 'renderItem',
    value: function renderItem(item, index) {
      if (item.type.name === 'Trigger' || item.props.type === 'trigger') {
        return _react2.default.createElement(
          _Trigger2.default,
          { key: index },
          item
        );
      } else if (item.type.name === 'hover' || item.props.type === 'hover') {
        return _react2.default.createElement(
          _Hover2.default,
          { key: index },
          item
        );
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var hoverComponentStyle = this.state.hoverComponentStyle;

      var childrenWithProps = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.props.children[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var child = _step.value;

          if (child.props) {
            if (child.type.name === 'Trigger' || child.props.type === 'trigger') {
              childrenWithProps.push(_react2.default.cloneElement(child, {
                setVisibility: this.setVisibility.bind(this),
                getCursorPos: this.getCursorPos.bind(this)
              }));
            } else if (child.type.name === 'Hover' || child.props.type === 'hover') {
              childrenWithProps.push(_react2.default.cloneElement(child, {
                styles: hoverComponentStyle
              }));
            }
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return _react2.default.createElement(
        'div',
        null,
        childrenWithProps.map(function (item, index) {
          return _this2.renderItem(item, index);
        })
      );
    }
  }, {
    key: 'setVisibility',
    value: function setVisibility(flag) {
      var hoverComponentStyle = this.state.hoverComponentStyle;

      var updatedStyles = null;
      if (flag) {
        updatedStyles = _extends({}, hoverComponentStyle, { display: 'block' });
      } else {
        updatedStyles = _extends({}, hoverComponentStyle, { display: 'none' });
      }
      this.setState({
        hoverComponentStyle: updatedStyles
      });
    }
  }, {
    key: 'getCursorPos',
    value: function getCursorPos(e) {
      var cursorX = e.pageX;
      var cursorY = e.pageY;
      var _props$options = this.props.options,
          followCursor = _props$options.followCursor,
          shiftX = _props$options.shiftX,
          shiftY = _props$options.shiftY;
      var hoverComponentStyle = this.state.hoverComponentStyle;

      var updatedStyles = null;
      if (!followCursor) {
        return;
      }
      if (isNaN(shiftX)) {
        shiftX = 0;
      }
      if (isNaN(shiftY)) {
        shiftY = 0;
      }
      updatedStyles = _extends({}, hoverComponentStyle, { top: cursorY + shiftY, left: cursorX + shiftX });
      this.setState({
        hoverComponentStyle: updatedStyles
      });
    }
  }]);

  return ReactHover;
}(_react.Component);

ReactHover.propTypes = {
  children: _propTypes2.default.array.isRequired,
  options: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string
};


ReactHover.Trigger = _Trigger2.default;
ReactHover.Hover = _Hover2.default;

exports.default = ReactHover;
