// isotrope/lib/index.js
exports.axes = ['x', 'y']
exports.sides = ['left', 'right', 'top', 'bottom']
exports.directions = ['left', 'right', 'up', 'down']
exports.dimensions = ['width', 'height']
exports.sideInDirection = {
  left: 'left',
  right: 'right',
  up: 'top',
  down: 'bottom',
}
exports.directionOf = {
  left: 'left',
  right: 'right',
  top: 'up',
  bottom: 'down',
}
exports.signumOf = {
  left: -1,
  right: 1,
  top: -1,
  bottom: 1,
  up: -1,
  down: 1,
}
exports.axisFor = {
  left: 'x',
  right: 'x',
  width: 'x',
  top: 'y',
  bottom: 'y',
  up: 'y',
  down: 'y',
  height: 'y',
}
exports.dimensionFor = {
  x: 'width',
  left: 'width',
  right: 'width',
  y: 'height',
  top: 'height',
  bottom: 'height',
  up: 'height',
  down: 'height',
}
exports.oppositeOf = {
  x: 'y',
  y: 'x',
  width: 'height',
  height: 'width',
  left: 'right',
  right: 'left',
  top: 'bottom',
  bottom: 'top',
  up: 'down',
  down: 'up',
}
exports.oppositeAxis = {
  x: 'y',
  y: 'x',
}
exports.oppositeDimension = {
  width: 'height',
  height: 'width',
}
exports.oppositeSide = {
  left: 'right',
  right: 'left',
  top: 'bottom',
  bottom: 'top',
}
exports.oppositeDirection = {
  left: 'right',
  right: 'left',
  up: 'down',
  down: 'up',
}
exports.loSide = {
  x: 'left',
  y: 'top',
}
exports.hiSide = {
  x: 'right',
  y: 'bottom',
}
exports.negativeDirection = {
  x: 'left',
  y: 'top',
}
exports.positiveDirection = {
  x: 'right',
  y: 'bottom',
}
